### PR TITLE
Add `onErrorRestart`

### PR DESCRIPTION
**Pull request by allantl**
Addresses #12, #13

### CODE REVIEW

**Comment by pjurczenko** [comment]
I'm wondering whether this would make sense to have complementary methods for fatal errors, i.e. `.onFatalErrorRestart` and `.onFatalErrorRestartIf`? This is just a thought, I'm just curious what you think about it 😉

**Comment by Avasil** [comment]
@pjurczenko 

> I'm wondering whether this would make sense to have complementary methods for fatal errors, i.e. `.onFatalErrorRestart` and `.onFatalErrorRestartIf`? This is just a thought, I'm just curious what you think about it 😉

I have answered in https://github.com/monix/monix-bio/pull/44#issuecomment-565714561

**Comment by allantl** [comment]
I've added the tests. Check the second commit.

File: core/shared/src/test/scala/monix/bio/TaskErrorSuite.scala [hunk 1]
```diff
@@ -243,87 +243,134 @@ object TaskErrorSuite extends BaseTestSuite {
...
+    val f = task.onErrorRestartIf(_ => tries < 5).runToFuture
+
+    assertEquals(f.value, Some(Failure(ex)))
+    assertEquals(tries, 1)
+  }
+
+  test("Task.onErrorRestartIf should restart on typed error") { implicit s =>
+    var tries = 0
+    val ex = DummyException("dummy")
+    val task = Task.eval { tries += 1; if (tries < 5) throw ex else 1 }.onErrorHandleWith {
```
Comment (Avasil) [inline core/shared/src/test/scala/monix/bio/TaskErrorSuite.scala]: We can also do something like `BIO.unit.flatMap { _ => tries += 1; if (p) BIO.raiseError("error") else 1 }` without throwing and handling `DummyException`

File: core/shared/src/test/scala/monix/bio/TaskErrorSuite.scala [hunk 2]
```diff
@@ -243,87 +243,134 @@ object TaskErrorSuite extends BaseTestSuite {
...
+    var tries = 0
+    val ex = DummyException("dummy")
+    val task = Task.eval { tries += 1; throw ex }.hideErrors.onErrorRestart(5)
+    val f = task.runToFuture
+
+    assertEquals(f.value, Some(Failure(ex)))
+    assertEquals(tries, 1)
+  }
+
+  test("Task.onErrorRestart should restart on typed error") { implicit s =>
```
Comment (Avasil) [inline core/shared/src/test/scala/monix/bio/TaskErrorSuite.scala]: I feel like we don't have to repeat tests for `raiseError(throwable)` and `raiseError(somethingelse)`. One of these should be enough

File: core/shared/src/test/scala/monix/bio/TaskErrorSuite.scala (hunk 1 above)
Comment (allantl) [reply core/shared/src/test/scala/monix/bio/TaskErrorSuite.scala]: Ah yes, this is much cleaner.

**Comment by allantl** [comment]
Ok, comments addressed.

**Comment by Avasil** [comment]
Thanks!

**Pull request merged**